File: 2047. Number of Valid Words in a Sentence/Solution.java
class Solution {
    public int countValidWords(String sentence) {
        String pattern = "([a-z]+-[a-z]+|[a-z]*)[!\\\\.,]?";
        int res = 0;
        for (String word : sentence.trim().split("\\s+")) {
            if (word.matches(pattern)) {
                res++;
            }
        }
        return res;
    }
}

File: 2047. Number of Valid Words in a Sentence/Solution.py
import re

class Solution:
    def countValidWords(self, sentence: str) -> int:
        pattern = re.compile(r'([a-z]+-[a-z]+|[a-z]*)[!.,]?')
        res = 0
        for word in sentence.split():
            if pattern.fullmatch(word):
                res += 1
        return res
